### ISSUE TITLE
Environment variables values are ignored when executing rules

### ISSUE THREAD

**Issue by chess-levin** [hadolint/hadolint#341]
- [?] This is a bug report
- [?] This is a feature request
- [ ] I searched existing issues before opening this one

### Expected behavior
hadloint throws an error every time the last USER statement in a dockerfile sets the user to root.

But when you make use of ENV variables to set the user like this:

ENV user=root
USER ${user}

hadolint doesn't find this issue.

I think this issue could not be solved only by an AST analysis. This may needs dynamic code analysis methods.

### Actual behavior
hadolint doesn't find that the last user set ist root

### Steps to reproduce the behavior

Set root user like this

ENV user=root
USER ${user}

**Comment by chagui** [comment]
Another example where this could be useful:

```Dockerfile
FROM python:3.9
ENV PIP_NO_CACHE_DIR="true"
RUN pip install pip~=20
```

```bash
$ docker run -i --rm hadolint/hadolint hadolint --version
Haskell Dockerfile Linter v1.19.0-44-gf364c91
$  docker run -i --rm hadolint/hadolint < Dockerfile
/dev/stdin:3 DL3042 Avoid use of cache directory with pip. Use `pip install --no-cache-dir <package>`
```

**Comment by Fixmetal** [comment]
Another example that could be useful:

```
FROM fixmetal/supercronic:v0.1.12-bullseye-slim-arm64

ARG GEOIPUPDATE_VER="4.9.0"
ARG GEOIPUPDATE="geoipupdate_${GEOIPUPDATE_VER}_linux_arm64.tar.gz"
ARG GEOIPUPDATE_URL="https://github.com/maxmind/geoipupdate/releases/download/v${GEOIPUPDATE_VER}/${GEOIPUPDATE}"

ADD ${GEOIPUPDATE_URL} /usr/local/bin
```

```
$ hadolint Dockerfile 
Dockerfile:8 DL3020 error: Use COPY instead of ADD for files and folders
```

It shouldn't complain at all because I'm actually using ADD to extract .tar.gz as per DL3020.

Hope this issue gets updated soon! Looking forward to this in a while.

**Comment by Fixmetal** [comment]
Hello guys,
is there any update on this?

**Comment by Mohataseem89** [comment]
hi any update??

i can fix the literal USER/ENV case in DL3002 as a self-contained first step would you want the variable-tracking logic extracted into a shared module so DL3042/DL3020 could reuse it later, or is a scoped single-rule fix fine for now?

**Comment by Mohataseem89** [comment]
After looking at the source, the issue is that `DL3002` only sees the raw text from `USER` (e.g. `${user}`) and never resolves it the parser does nott expand variables so the rule has to do it 

My plan:
- Track `ENV` key/value pairs in `DL3002`'s accumulator, reset on each `FROM`
- Resolve `$VAR` / `${VAR}` references in the `USER` instruction against those pairs before the root check
- Leave unresolved references untouched (no false positives if the variable was never set via `ENV`)

scope is intentionally limited to simple substitution no full shell expansion (no `${VAR:-default}` etc.) since that is what the literal repro cases need and keeps the change self-contained.

one question before I open the PR: would you prefer this variable-tracking logic live inside `DL3002.hs` only or extracted into a shared helper so other rules (e.g. DL3042) could reuse it later?
